src/toil/provisioners/abstractProvisioner.py

```diff
@@ -234,189 +395,620 @@ def destroyCluster(self):
+                --path.procfs /host/proc \\
+                --path.sysfs /host/sys \\
+                --collector.filesystem.ignored-mount-points ^/(sys|proc|dev|host|etc)($|/)
+            '''))
+        
+    def addToilService(self, config: InstanceConfiguration, role: str, keyPath: str = None, preemptable: bool = False):
+        """
+        Add the Toil leader or worker service to an instance configuration.
+        
+        Will run Mesos master or agent as appropriate in Mesos clusters.
+        For Kubernetes clusters, will just sleep to provide a place to shell
+        into on the leader, and shouldn't run on the worker.
+        
+        :param role: Should be 'leader' or 'worker'. Will not work for 'worker' until leader credentials have been collected.
+        :param keyPath: path on the node to a server-side encryption key that will be added to the node after it starts. The service will wait until the key is present before starting.
+        :param preemptable: Whether a woeker should identify itself as preemptable or not to the scheduler.
+        """
+        
+        # If keys are rsynced, then the mesos-agent needs to be started after the keys have been
+        # transferred. The waitForKey.sh script loops on the new VM until it finds the keyPath file, then it starts the
+        # mesos-agent. If there are multiple keys to be transferred, then the last one to be transferred must be
+        # set to keyPath.
+        MESOS_LOG_DIR = '--log_dir=/var/lib/mesos '
+        LEADER_DOCKER_ARGS = '--registry=in_memory --cluster={name}'
+        # --no-systemd_enable_support is necessary in Ubuntu 16.04 (otherwise,
+        # Mesos attempts to contact systemd but can't find its run file)
+        WORKER_DOCKER_ARGS = '--work_dir=/var/lib/mesos --master={ip}:5050 --attributes=preemptable:{preemptable} --no-hostname_lookup --no-systemd_enable_support'
+
+        if self.clusterType == 'mesos':
+            if role == 'leader':
+                entryPoint = 'mesos-master'
+                entryPointArgs = MESOS_LOG_DIR + LEADER_DOCKER_ARGS.format(name=self.clusterName)
+            elif role == 'worker':
+                entryPoint = 'mesos-agent'
+                entryPointArgs = MESOS_LOG_DIR + WORKER_DOCKER_ARGS.format(ip=self._leaderPrivateIP,
+                                                            preemptable=preemptable)
+            else:
+                raise RuntimeError("Unknown role %s" % role)
+        elif self.clusterType == 'kubernetes':
+            if role == 'leader':
+                entryPoint = 'sleep'
+                entryPointArgs = 'infinity'
+            else:
+                raise RuntimeError('Toil service not needed for %s nodes in a %s cluster',
+                    role, self.clusterType)
         else:
-            raise RuntimeError("Unknown role %s" % role)
+            raise RuntimeError('Toil service not needed in a %s cluster', self.clusterType)
```

Comment:
Yes. If we get here we've forgotten to implement a cluster type, or are trying to add a service to a cluster type that shouldn't have it. I guess we could ignore it and make this method responsible for determining if the service is needed, instead of the caller, but I don't see why we would warn.